packages/lodestar/src/sync/backfill/verify.ts

```diff
@@ -0,0 +1,37 @@
+import {allForks, CachedBeaconState} from "@chainsafe/lodestar-beacon-state-transition";
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {SignedBeaconBlock} from "@chainsafe/lodestar-types/lib/allForks";
+import {Root} from "../../../../types/lib";
+import {IBlsVerifier} from "../../chain/bls";
+import {BackfillSyncError, BackfillSyncErrorCode} from "./errors";
+
+export async function verifyBlocks(
+  config: IBeaconConfig,
+  bls: IBlsVerifier,
+  state: CachedBeaconState<allForks.BeaconState>,
+  blocks: SignedBeaconBlock[],
+  anchorRoot: Root
+): Promise<void> {
+  if (blocks.length === 0) {
+    return;
+  }
+  const nextRoot: Root = anchorRoot;
+  const signatures: ReturnType<typeof allForks["getProposerSignatureSet"]>[] = [];
+  for (const block of blocks.reverse()) {
+    if (
+      !config.types.Root.equals(
+        config.getForkTypes(block.message.slot).BeaconBlock.hashTreeRoot(block.message),
+        nextRoot
+      )
+    ) {
+      if (config.types.Root.equals(nextRoot, anchorRoot)) {
+        throw new BackfillSyncError({code: BackfillSyncErrorCode.NOT_ANCHORED});
+      }
+      throw new BackfillSyncError({code: BackfillSyncErrorCode.NOT_LINEAR});
+    }
+    signatures.push(allForks.getProposerSignatureSet(state, block));
```

Comment:
Should we gather the signatures after this for loop in a map? No need to do `ReturnType<typeof allForks["getProposerSignatureSet"]>[]` then

Reply:
but bls.verifySignatureSets expects array

Reply:
```ts
const signatures = blocks.map(block => allForks.getProposerSignatureSet(state, block))
if (!(await bls.verifySignatureSets(signatures))) {
    throw new BackfillSyncError({code: BackfillSyncErrorCode.INVALID_SIGNATURE});
  }
```